doc/design/api.md

```diff
@@ -0,0 +1,277 @@
+# PaddlePaddle Design Doc
+
+## Ingredients
+
+As the first step of our design, we list important concepts in deep
+learning and try to figure their relationship, as shown below:
+
+```
+Model = {topology, parameters}
+
+Evaluator = {Model*, activations}
+- forward
+- test(cost, ...)
```

Comment:
For the "Is `Evaluator` a necessary concept for Paddle?". I think we need some object for caching intermediate tensors. From what I understand (asked this question in yesterday's meeting), paddle don't release intermediate tensors after a forward pass even only for inference. I think that is reasonable. Since `malloc` and `free` takes time. If we don't release intermediate tensors, but reusing them by overriding, we can save forward time. (at least, release or not could be tunable argument).
So if we don't release, we need an object to save the states(all intermediate tensors). `Evaluator` could be such an object.